python/tvm/relay/build_module.py

```diff
@@ -217,30 +229,147 @@ def build(func,
     target = target if target else _target.current_target()
     if target is None:
         raise ValueError("Target is not set in env or passed as argument.")
-    target = _target.create(target)
+
+    if isinstance(target, dict):
+        target, fallback_device = \
+                _update_heterogeneous_inputs(target, fallback_device)
+    elif isinstance(target, (str, _target.Target)):
+        target = _target.create(target)
+    else:
+        raise ValueError("target must be the type of str, tvm.target.Target," +
+                         "or dict of device name to target")
 
     # If current dispatch context is fallback context (the default root context),
     # then load pre-tuned parameters from TopHub
     if isinstance(autotvm.DispatchContext.current, autotvm.FallbackContext):
-        tophub_context = autotvm.tophub.context(target)
+        if isinstance(target, dict):
+            tophub_context = autotvm.tophub.context(list(target.values()))
+        else:
+            tophub_context = autotvm.tophub.context(target)
     else:
         tophub_context = autotvm.util.EmptyContext()
 
     cfg = BuildConfig.current
 
     with tophub_context:
         func = optimize(func, target, params)
+        # Annotate the ops for heterogeneous execution.
+        if isinstance(target, dict):
+            func, target = _run_device_annotation_passes(func, target,
+                                                         fallback_device)
         # Fuse ops before running code gen
         func = ir_pass.infer_type(func)
         func = ir_pass.fuse_ops(func, cfg.opt_level)
         # Graph code generation
         func = ir_pass.infer_type(func)
         graph_gen = _graph_gen.GraphRuntimeCodegen(mod=None, target=target)
         graph_json, lowered_funcs, params = graph_gen.codegen(func)
-        mod = _tvm_build_module(lowered_funcs, target=target, target_host=target_host)
+        mod = _tvm_build_module(
+            lowered_funcs, target=target, target_host=target_host)
     return graph_json, mod, params
 
 
+def _update_heterogeneous_inputs(target, fallback_device=None):
+    """Update the target and fallback device required for heterogeneous
+    compilation. CPU is used as the fallback device if it wasn't provided.
+    Meanwhile, a CPU device id and "llvm" pair will be added to the target
+    dictionary in this case.
+
+    Parameters
+    ----------
+    target : dict of str(i.e. device/context name) to str/tvm.target.Target.
```

Comment:
can key be `TVMContext`?

Reply:
It could, but I used also for homogeneous execution by making the key `0`. So it might be more convenient to let it be `int`

Reply:
@yzhliu Yes. They need to because it is more user friendly. This is just the internal representation.